torch/futures/__init__.py

```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
     Wrapper around a ``torch._C.Future`` which encapsulates an asynchronous
     execution of a callable, e.g. :meth:`~torch.distributed.rpc.rpc_async`. It
     also exposes a set of APIs to add callback functions and set results.
+
+    .. warning:: GPU support is a beta feature, subject to changes.
     """
 
     def done(self) -> bool:
         r"""
         Return ``True`` if this ``Future`` is done. A ``Future`` is done if it
         has a result or an exception.
+
+        If the value contains tensors that reside on GPUs, ``Future.done()``
+        will return ``True`` even if the asynchronous kernels that are
+        populating those tensors haven't yet completed running on the device,
+        because at such stage the result is already usable, provided one
+        performs the appropriate synchronizations (see :meth:`wait`).
         """
         return super().done()
 
     def wait(self, non_blocking: bool = False) -> T:
         r"""
         Block until the value of this ``Future`` is ready.
 
+        If the value contains tensors that reside on GPUs, then an additional
+        synchronization is performed with the kernels (executing on the device)
+        which may be asynchronously populating those tensors. Such sync can be:
```

Comment:
I don't think this is true for all futures we return in PyTorch. As per my understanding only ProcessGroupNCCL and our RPC framework guarantee this, but do ops like `torch.jit.fork` respect this? For example, I can do this today with GPU tensors:

```
import torch

@torch.jit.script
def add_one(inp):
    return inp + 1

t = torch.rand(3,3).cuda()
fut = torch.jit.fork(add_one, t)
print(fut.wait())
```

I believe the Future returned by torch.jit.fork will be a CPU future and not a CUDA future despite containing GPU tensors. The non_blocking behavior is controlled by what kind of Future you return and not what kind of tensors the Future holds. Maybe we should update the docs to reflect which APIs would return a CUDA future and the non_blocking flag applies only to those futures.

Reply:
Not sure if it changed recently, but today it seems that `torch.jit.fork()` just calls `torch._C.fork()`:
https://github.com/pytorch/pytorch/blob/43f6deb6e447aae9d2edbd9441043b03c6ff0992/torch/jit/_async.py#L82

Which in turn is defined as returning a `torch._C.Future`:
https://github.com/pytorch/pytorch/blob/43f6deb6e447aae9d2edbd9441043b03c6ff0992/torch/_C/__init__.pyi.in#L168

Which is (basically) the same class as `torch.futures.Future`.